src/persist-client/src/internal/state.rs

```diff
@@ -336,12 +369,36 @@ where
         Continue(Since(reader_current_since))
     }
 
-    pub fn heartbeat_reader(
+    pub fn compare_and_downgrade_since(
+        &mut self,
+        reader_id: &CriticalReaderId,
+        (expected_opaque, expected_since): (Opaque, &Antichain<T>),
+        (new_opaque, new_since): (Opaque, &Antichain<T>),
+    ) -> ControlFlow<Infallible, Result<(Opaque, Since<T>), (Opaque, Since<T>)>> {
+        let reader_state = self.critical_reader(reader_id);
+
+        if &reader_state.since == expected_since && reader_state.opaque == expected_opaque {
+            assert!(PartialOrder::less_than(expected_since, new_since));
+            reader_state.since = new_since.clone();
+            reader_state.opaque = new_opaque.clone();
+            self.update_since();
+            Continue(Ok((new_opaque, Since(new_since.clone()))))
+        } else {
```

Comment:
still undecided if we should try to/need to detect if this handle has already applied this specific change (but is retrying due to indeterminate error).

if we did it, the branch would be something like:

```
else if &reader_state.since == new_since && reader_state.opaque == new_opaque {
    // break and tell the caller we were the one to apply this change
}
```

@danhhz any thoughts here?

Reply:
I think this is something we need to punt back to the users of this method

Reply:
👍 that should already happen, and they'd get an `Err` potentially with the since/opaque value they passed in. In the use cases discussed so far, I think this would be handled just fine (when fencing, the `opaque` val matters but not the `since`, and when not fencing, the `Err` case is, I think, ignorable)

Reply:
oh, sorry! I meant "the decision on what the behavior should be here" should be something we let the users of it decide :D

Reply:
hah sure -- cc @bkirwi @petrosagg @frankmcsherry 

the issue here is: if we tried to `compare_and_downgrade_since` on a critical handle, we receive an indeterminate error from Cockroach (we don't know if our write succeeded or not), and then the latest state has advanced such that the expected `(since, opaque)` no longer match, what's the most intuitive answer to return to the caller?

Options:
1. An indeterminate error
1. The current `(since, opaque)` value in an `Err`. These values may be what the request initially asked for, but it'll be returned in an `Err` because we never got confirmation _our_ request succeeded (maybe a concurrent `SinceHandle` to this same critical reader ID wrote the same state).
1. If `(requested_since, requested_opaque) == (current_since, current_opaque)` we return an `Ok`, otherwise we return the latest state in an `Err` as above. If the caller has a proper fencing token and sets their `opaque` value to that token, I think it is safe to assume this state can only be reached if our handle's write did succeed. If `opaque` is some shared value (i.e. it's safe for multiple `SinceHandle`s to work against the same critical reader id at the same time), there is a chance we could return `Ok` if another handle performed an identical operation, its operation succeeded, and ours did not.

Hopefully that makes sense... it's a bit of a doozy to type it all out.

Reply:
It strikes me that 3 is probably the simplest of the options, and also totally compatible with the use-cases we've discussed so far. (If we just want to join the old and new values, great; if we're fencing with an in-process epoch, presumably if we're racing with anyone it's our own previous call.) 2 seems manageable since the client can do the filtering; 1 seems harder to deal with and nice to avoid if we can!

Reply:
👍

Dan had a nice suggestion for the return type / guarantees, and I think it also tidies this one up more in line with (3) than the others. We return an `Ok` if the latest state has `!since.less_than(new_since) && state.opaque == expected_opaque`, so either we see the result of the downgrade exactly, or something more recent. We only `Err` if the opaque value has changed from what we expect